baseplate.go

```diff
@@ -0,0 +1,245 @@
+package baseplate
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+	"os"
+	"time"
+
+	yaml "gopkg.in/yaml.v2"
+
+	"github.com/reddit/baseplate.go/batcherror"
+	"github.com/reddit/baseplate.go/edgecontext"
+	"github.com/reddit/baseplate.go/log"
+	"github.com/reddit/baseplate.go/metricsbp"
+	"github.com/reddit/baseplate.go/runtimebp"
+	"github.com/reddit/baseplate.go/secrets"
+	"github.com/reddit/baseplate.go/tracing"
+)
+
+// Config is a general purpose config for assembling a Baseplate server
+type Config struct {
+	// Addr is the local address to run your server on.
+	//
+	// It should be in the format "${IP}:${Port}", "localhost:${Port}",
+	// or simply ":${Port}".
+	Addr string
+
+	// Timeout is the socket connection timeout for Servers that
+	// support that.
+	Timeout time.Duration
+
+	// StopTimeout is the timeout for the Stop command for the service.
+	//
+	// If this is not set, then no timeout will be set on the Stop command.
+	StopTimeout time.Duration `yaml:"stopTimeout"`
+
+	Log     log.Config
+	Metrics metricsbp.Config
+	Secrets secrets.Config
+	Sentry  log.SentryConfig
+	Tracing tracing.Config
+}
+
+// Baseplate is the general purpose object that you build a Server on.
+type Baseplate interface {
+	io.Closer
+
+	Config() Config
+	EdgeContextImpl() *edgecontext.Impl
+	Secrets() *secrets.Store
+}
+
+// Server is the primary interface for baseplate servers.
+type Server interface {
```

Comment:
Why is this an interface?

Reply:
Because we have methods in baseplate that operate on a `Server` (specifically `baseplate.Serve`) and have multiple implementations.  I think also _generally_ for something like this an interface is good for just communicating what is the expected API.

Reply:
Fair enough, I thought the only implementation is the test server.